api/src/main/java/bio/terra/tanagra/service/search/Filter.java

```diff
@@ -74,5 +76,67 @@ public static ArrayFunction create(List<Filter> operands, Operator operator) {
     }
   }
 
-  // TODO implement a relationship filter tht allows chaining to another entity.
+  /**
+   * A {@link Filter} that introduces a constraint that there must be a relationship to another
+   * entity.
+   *
+   * <p>This introduces a new {@link Variable} and implicitly binds it to an entity in {@link
+   * #boundAttribute()}. The bound and related entities should have a relationship across the bound
+   * attributes. The outer variable may be any variable introduced higher in the outer filter scope.
+   *
+   * <p>This is equivalent to the pseudo-SQL ":outerAttribute IN (SELECT :boundAttribute FROM
+   * :boundAttribute.entity AS :boundAttribute.variable WHERE :filter)".
+   *
+   * <p>Or in words ∃x∈boundEntity( x.boundAttribute = outerAttribute ∧ filter): there exists an
+   * instance 'x' of {@link #boundAttribute()}'s Entity where x's {@link #boundAttribute()} equals
+   * {@link #outerAttribute()} and {@link #filter}.
+   */
+  @AutoValue
+  abstract class RelationshipFilter implements Filter {
+
+    /**
+     * The attribute of the newly introduced variable's entity that is equal to {@link
+     * #outerAttribute}.
+     */
+    public abstract AttributeVariable boundAttribute();
+
+    /** The attribute of an outer variable that's equal to {@link #boundAttribute}. */
+    public abstract AttributeVariable outerAttribute();
+
+    /** The filter to apply to the bound variable. This may refer to outer variables. */
+    public abstract Filter filter();
+
+    @Override
+    public <R> R accept(Visitor<R> visitor) {
+      return visitor.visitRelationship(this);
+    }
+
+    public static Builder builder() {
+      return new AutoValue_Filter_RelationshipFilter.Builder();
+    }
+
+    /** Builder for {@link RelationshipFilter}. */
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      public abstract Builder boundAttribute(AttributeVariable boundAttribute);
+
+      public abstract AttributeVariable boundAttribute();
+
+      public abstract Builder outerAttribute(AttributeVariable outerAttribute);
+
+      public abstract AttributeVariable outerAttribute();
+
+      public abstract Builder filter(Filter filter);
+
+      public RelationshipFilter build() {
+        Preconditions.checkArgument(
+            !boundAttribute().variable().equals(outerAttribute().variable()),
+            "Cannot bind a new variable to the outer attribute's variable. ");
```

Comment:
If I got this error, I might be stuck. Is there any guidance that can be provided in this exception as to how the user should correct their usage?

also nit: trailing space

Reply:
thanks, done.